sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -833,6 +897,124 @@ void Scheduler::GraphBuilder::removeRecordForMemObj(SYCLMemObjI *MemObject) {
+
+  if (Command *DepCmd = reinterpret_cast<Command *>(DepEvent->getCommand()))
+    DepCmd->addUser(ConnectCmd);
+
+  EmptyCommand *EmptyCmd = nullptr;
+
+  if (Dep.MDepRequirement) {
+    Requirement *Req = const_cast<Requirement *>(Dep.MDepRequirement);
+
+    // make ConnectCmd depend on requirement
+    {
+      MemObjRecord *Record = getMemObjRecord(Req->MSYCLMemObj);
+      Dep.MDepCommand->addUser(ConnectCmd);
+
+      AllocaCommandBase *AllocaCmd =
+          findAllocaForReq(Record, Req, DepEventContext);
+      assert(AllocaCmd && "There must be alloca for requirement!");
+
+      std::set<Command *> Deps = findDepsForReq(Record, Req, DepEventContext);
+      assert(Deps.size() && "There must be some deps");
+
+      for (Command *ReqDepCmd : Deps) {
+        // we don't want to depend on any host task as the only "entry point" to
+        // host task is its empty cmd which is in Deps anyway
+        if (ReqDepCmd->getType() == Command::CommandType::RUN_CG) {
+          auto *Cmd = static_cast<ExecCGCommand *>(ReqDepCmd);
+          if (Cmd->getCG().getType() == CG::CGTYPE::CODEPLAY_HOST_TASK)
+            continue;
+        }
+
+        ConnectCmd->addDep(DepDesc{ReqDepCmd, Req, AllocaCmd});
+        ReqDepCmd->addUser(ConnectCmd);
+      }
+
+      updateLeaves(Deps, Record, Req->MAccessMode);
+      addNodeToLeaves(Record, ConnectCmd, Req->MAccessMode);
+    }
+
+    const auto &Reqs = std::vector<Requirement *>(1, Req);
+    EmptyCmd = addEmptyCmd(ConnectCmd, Reqs,
+                           Scheduler::getInstance().getDefaultHostQueue(),
+                           Command::BlockReason::HostTask);
+    // Dependencies for EmptyCmd are set in addEmptyCmd for provided Reqs.
+
+    // Depend Cmd on empty command
+    {
+      DepDesc CmdDep = Dep;
+      CmdDep.MDepCommand = EmptyCmd;
+
+      Cmd->addDep(CmdDep);
+    }
+  } else {
+    EmptyCmd = addEmptyCmd(ConnectCmd, {},
+                           Scheduler::getInstance().getDefaultHostQueue(),
+                           Command::BlockReason::HostTask);
+
+    // There is no requirement thus, empty command will only depend on
+    // ConnectCmd via its event.
+    EmptyCmd->addDep(ConnectCmd->getEvent());
+    ConnectCmd->addDep(DepEvent);
+
+    // Depend Cmd on empty command
+    Cmd->addDep(EmptyCmd->getEvent());
+  }
+
+  EmptyCmd->addUser(Cmd);
+
+  ConnectCmd->MEmptyCmd = EmptyCmd;
+
+  // FIXME graph builder shouldn't really enqueue commands. We're in the middle
+  // of enqueue process for some command Cmd. We're going to add a dependency
+  // for it. Need some nice and cute solution to enqueue ConnectCmd via standard
+  // scheduler/graph processor mechanisms.
+  EnqueueResultT Res;
+  bool Enqueued = Scheduler::GraphProcessor::enqueueCommand(ConnectCmd, Res);
+  if (!Enqueued && EnqueueResultT::SyclEnqueueFailed == Res.MResult)
+    throw runtime_error("Failed to enqueue a sync event between two contexts",
+                        PI_INVALID_OPERATION);
+}
```

Comment:
need to handle when Enqueued is false but the error is not MResult. Maybe an assert is enough?

Reply:
This assertion should go in distinct patch as it requires modifications in other places where enqueue is done.